src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+
+            try
+            {
+                ReadOnlySpan<char> passwordChars = password.DangerousGetSpan();
+
+                if (_pfxAsn.MacData.HasValue)
+                {
+                    VerifyAndDecrypt(passwordChars, authSafeContents);
+                }
+                else if (passwordChars.IsEmpty)
+                {
+                    try
+                    {
+                        // Try the empty password first.
+                        // If anything goes wrong, try the null password.
+                        //
+                        // The same password has to work for the entirety of the file,
+                        // null and empty aren't interchangeable between parts.
+                        Decrypt("", authSafeContents);
+                    }
+                    catch (CryptographicException)
+                    {
+                        ContentInfoAsn[] partialSuccess = _safeContentsValues;
+                        _safeContentsValues = null;
+
+                        if (partialSuccess != null)
+                        {
+                            ArrayPool<ContentInfoAsn>.Shared.Return(partialSuccess);
+                        }
+
+                        Decrypt(null, authSafeContents);
+                    }
+                }
+                else
+                {
+                    Decrypt(passwordChars, authSafeContents);
+                }
+            }
+            catch (Exception e)
+            {
+                throw new CryptographicException(SR.Cryptography_Pfx_BadPassword, e);
+            }
+            finally
+            {
+                password.DangerousRelease();
+            }
+        }
+
+        private void VerifyAndDecrypt(ReadOnlySpan<char> password, ReadOnlyMemory<byte> authSafeContents)
+        {
+            Debug.Assert(_pfxAsn.MacData.HasValue);
+            ReadOnlySpan<byte> authSafeSpan = authSafeContents.Span;
+
+            if (password.Length == 0)
+            {
+                // VerifyMac produces different answers for the empty string and the null string,
+                // when the length is 0 try empty first (more common), then null.
+                if (_pfxAsn.VerifyMac("", authSafeSpan))
+                {
+                    Decrypt("", authSafeContents);
+                    return;
+                }
+
+                if (_pfxAsn.VerifyMac(default, authSafeSpan))
+                {
+                    Decrypt(default, authSafeContents);
+                    return;
+                }
+            }
+            else if (_pfxAsn.VerifyMac(password, authSafeSpan))
+            {
+                Decrypt(password, authSafeContents);
+                return;
+            }
+
+            throw new CryptographicException(SR.Cryptography_Pfx_BadPassword);
+        }
+
+        private void Decrypt(ReadOnlySpan<char> password, ReadOnlyMemory<byte> authSafeContents)
+        {
+            if (_safeContentsValues == null)
```

Comment:
any chance this method might be split a bit, it's a bit hard to follow what's going on in here

Reply:
Broke out 4 methods

* DecodeSafeContents
* DecryptAndProcessSafeContents
* ExtractPrivateKeys
* BuildCertsWithKeys

now this method is just the scaffolding to get from one state to the next.